libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * @file Inliner.cpp
+ * Inlines small code snippets by replacing JUMP with a copy of the code jumped to.
+ */
+
+#include <libevmasm/Inliner.h>
+
+#include <libevmasm/AssemblyItem.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/slice.hpp>
+
+#include <optional>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::evmasm;
+
+bool Inliner::isInlineCandidate(u256 const& _tag, InlinableBlock const& _block) const
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+
+	// Never inline tags that reference themselves.
+	for (AssemblyItem const& item: _block.items)
+		if (item.type() == PushTag)
+			if (_tag == item.data())
+				return false;
+
+	return true;
+}
+
+map<u256, Inliner::InlinableBlock> Inliner::determineInlinableBlocks(AssemblyItems const& _items) const
+{
+	std::map<u256, ranges::span<AssemblyItem const>> inlinableBlockItems;
+	std::map<u256, uint64_t> numPushTags;
+	std::optional<size_t> lastTag;
+	for (auto&& [index, item]: _items | ranges::views::enumerate)
+	{
+		// The number of PushTag's approximates the number of calls to a block.
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
+		{
+			if (item == Instruction::JUMP)
+				inlinableBlockItems.emplace(
+					piecewise_construct,
+					forward_as_tuple(_items[*lastTag].data()),
+					forward_as_tuple(_items | ranges::views::slice(*lastTag + 1, index + 1))
+				);
```

Comment:
Why is this so complicated?
```suggestion
				inlinableBlockItems[_items[*lastTag].data()]= _items | ranges::views::slice(*lastTag + 1, index + 1);
```

Reply:
Because if I do it often enough, maybe people will get used to it as the canonical way for zero-cost in-place construction of map elements, which it is ;-P.

Reply:
They won't ;)

Reply:
Well, ok, I'm changing it... but I still must say: you are aware that this construct *always* works and is *always* optimal, and I'd say once you have seen it only once, it's also absolutely clear what it does...
Whereas the (IMHO only seemingly) simpler ``...[...] = ...;`` has to construct and - if you're unlucky - copy two temporaries, even before that default-construct an empty element to be copied (or in the best case moved) over, which may not even be possible, if you don't have a default-constructor, it gives no feedback about it already having been there or not, etc.

So I think one can very well argue that ``...[...] = ...`` is *always* inferior - if both work, in-place constructing is always cheaper, and there are cases in which *only* inplace-constructing will even work.

But yeah, on the other hand, it's not important enough to argue, so having said all that, I'll still just change it ;-).

Reply:
It's not inferior, it is so much better with regards to readability. And reduced readability can easily lead to bugs. I would always prefer a language with syntax over a language without syntax.
```(for (def i) (gt i 0) (assign i (add i 1)) (mstore i (mul i 2)))```
Which elements of the for loop belong where?

Reply:
I don't think that's analog in any way - ``container.emplace(piecewise_construct, `` exactly says what it does - it piecewise construct something into my container. And ``forward_as_tuple(keyArgs...)`` and ``forward_as_tuple(valueArgs...)`` exactly say how the stuff is constructed - using forwarded arguments.
On the other hand a whole bunch of hidden things happen in ``container[keyArgs] = valueArgs;``. First I construct a temporary key object - then I potentially default-construct a value object in the container, then I construct a temporary value type, then I move or copy that value type into the container. Loads of stuff no-one might ever suspect. And every step along the way may be undesirable or may just not work...

Reply:
But well, I did change it, so not much point arguing about it ;-).